src/psyclone/gocean1p0.py

```diff
@@ -404,86 +411,194 @@ def __init__(self, parent=None,
-                    {'inner': {'start': "1", 'stop': "{stop}"},
-                     'outer': {'start': "1", 'stop': "{stop}"}}
+                GOLoop._bounds_lookup['offset_any'][gridpt_type][itspace] = \
+                    {'inner': {'start': "{start}-1", 'stop': "{stop}"},
+                     'outer': {'start': "{start}-1", 'stop': "{stop}"}}
         # For 'every' grid-point type
         for offset in SUPPORTED_OFFSETS:
             for itspace in VALID_ITERATES_OVER:
-                self._bounds_lookup[offset]['every'][itspace] = \
-                    {'inner': {'start': "1", 'stop': "{stop}+1"},
-                     'outer': {'start': "1", 'stop': "{stop}+1"}}
+                GOLoop._bounds_lookup[offset]['every'][itspace] = \
+                    {'inner': {'start': "{start}-1", 'stop': "{stop}+1"},
+                     'outer': {'start': "{start}-1", 'stop': "{stop}+1"}}
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def add_bounds(bound_info):
+        '''
+        Adds a new iteration space to PSyclone. An iteration space in the
+        gocean1.0 API is for a certain offset type and field type. It defines
+        the loop boundaries for the outer and inner loop. The format is a
+        ":" separated tuple:
+        bound_info = offset-type:field-type:iteration-space:outer-start:
+                      outer-stop:inner-start:inner-stop
+        Example:
+        bound_info = offset_ne:ct:all_pts:{start}-1:{stop}+1:{start}:{stop}
+
+        The expressions {start} and {stop} will be replaced with the loop
+        indices that correspond to the inner points (i.e. non-halo or
+        boundary points) of the field. So the index {start}-1 is actually
+        on the halo / boundary.
+
+        :param str bound_info: A string that contains a ":" separated \
+               tuple with the iteration space definition.
+        :raises ValueError: if bound_info is not a string.
+        :raises ConfigurationError: if bound_info is not formatted correctly.
+        '''
 
+        if not isinstance(bound_info, str):
+            raise InternalError("The parameter 'bound_info' must be a string, "
+                                "got '{0}' (type {1}".format(bound_info,
+                                                             type(bound_info)))
+
+        data = bound_info.split(":")
+        if len(data) != 7:
+            from psyclone.configuration import ConfigurationError
+            raise ConfigurationError("An iteration space must be in the form "
+                                     "\"offset-type:field-type:"
+                                     "iteration-space:outer-start:"
+                                     "outer-stop:inner-start:inner-stop\"\n"
+                                     "But got \"{0}\"".format(bound_info))
+
+        if len(GOLoop._bounds_lookup) == 0:
+            GOLoop.setup_bounds()
```

Comment:
Strange, for me it is reported to be covered. I modified gocen1p0_config_test.py so that only the first test would run, the one which includes:
```
   # Check that --config with a parameter is accepted
    main(options+["--config", config_file, f90_file])
```
joerg@localhost:~/work/psyclone/src/psyclone/tests> py.test  --cov-report=html  --cov=psyclone.gocean1p0  --cov-report term-missing  ./gocean1p0_config_test.py 
========================================================================================== test session starts ===========================================================================================
platform linux2 -- Python 2.7.14, pytest-3.6.1, py-1.5.3, pluggy-0.6.0
rootdir: /home/joerg/work/psyclone, inifile:
plugins: xdist-1.23.0, pylint-0.7.1, pep257-0.0.5, pep8-1.0.6, forked-0.2, flakes-2.0.0, cov-2.5.1
collected 1 item                                                                                                                                                                                         

gocean1p0_config_test.py .                                                                                                                                                                         [100%]

---------- coverage: platform linux2, python 2.7.14-final-0 ----------
Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
/home/joerg/work/psyclone/src/psyclone/gocean1p0.py     535    157    71%   144, 164, 194, 218, 231, 265-268, 271-274, 318-322, 326-331, 337-338, 350, 364, 378, 411, 415, 520, 526-527, 545-546, 568-570, 580, 584, 633-664, 707-728, 734-744, 753, 760, 765, 774, 796, 841, 862, 881, 908, 915, 941, 956, 971, 993-1017, 1028-1029, 1041-1042, 1076, 1112, 1122, 1194, 1206-1263, 1268, 1286-1287, 1302-1303, 1313-1314, 1342-1348, 1391, 1406-1413, 1421, 1430, 1452, 1464, 1469, 1477, 1482, 1493, 1515, 1555-1564

This does not list line 533 or 534 as uncovered :( I didn't check if this is correct - but it again raises the question: which version are you using? Maybe I need to update, or change the parameters I am using for testing?

Reply:
That is strange. I (almost) always work on what codecov says as that way I don't have to do it myself :-) Click on 'Continue to review full report at Codecov.` in the codecov comment on the PR and go to the `Diff` tab.

Reply:
OK, I think I found the reason for it: my command line runs only a few of the tests (e.g. only the gocean1p0 and the config_test - but when I run all tests I get the same failure (due to the runtime on my VM I usually don't do that). OK, I'll figure out how to reset the test in gocean1p0_config_test.py appropriately so that it will always test that line.